examples/android/clientcache/app/src/main/java/io/grpc/clientcacheexample/SafeMethodCachingInterceptor.java

```diff
@@ -0,0 +1,297 @@
+package io.grpc.clientcacheexample;
+
+import android.util.LruCache;
+import com.google.common.base.Splitter;
+import com.google.protobuf.MessageLite;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.Deadline;
+import io.grpc.ForwardingClientCall;
+import io.grpc.ForwardingClientCallListener;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * An example of an on-device cache for Android implemented using the {@link ClientInterceptor} API.
+ *
+ * <p>Client-side cache-control directives are not directly supported. Instead, two call options can
+ * be added to the call: no-cache (always go to the network) or only-if-cached (never use network;
+ * if response is not in cache, the request fails).
+ *
+ * <p>This interceptor respects the cache-control directives in the server's response: max-age
+ * determines when the cache entry goes stale. no-cache, no-store, and no-transform entirely skip
+ * caching of the response. must-revalidate is ignored, as the cache does not support returning
+ * stale responses.
+ *
+ * <p>Note: other response headers besides cache-control (such as Expiration, Varies) are ignored by
+ * this implementation.
+ */
+final class SafeMethodCachingInterceptor implements ClientInterceptor {
+  public static CallOptions.Key<Boolean> NO_CACHE_CALL_OPTION =
+      CallOptions.Key.of("no-cache", false);
+  public static CallOptions.Key<Boolean> ONLY_IF_CACHED_CALL_OPTION =
+      CallOptions.Key.of("only-if-cached", false);
+
+  public static final class Key {
+    private final String fullMethodName;
+    private final MessageLite request;
+
+    public Key(String fullMethodName, MessageLite request) {
```

Comment:
Why not take a MethodDescriptor, and parameterize on the request type from the MD?

Reply:
Would be nicer, but MethodDescriptor doesn't implement equals and reference equality isn't sufficient here. "Current practice" (e.g., in this PR and our tests) allow copying the MethodDescriptor and setting it to safe as a valid means of turning on GET. Added test for this case that fails if we key on MD by reference equality.